src/HttpRequest.h

```diff
@@ -105,6 +105,10 @@ class HttpRequest: public Http::Message
     /// associates the request with a from-client connection manager
     void manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntryPointer &al);
 
+    /// whether the master transaction traffic is intercepted
+    /// see TrafficMode::isIntercepted() for details
+    bool interceptedXaction() const { assert(masterXaction && masterXaction->squidPort); return masterXaction->squidPort->flags.isIntercepted(); }
```

Comment:
MasterXaction::squidPort should exist here - I fixed several places in 0e27722. I assume squidPort should be non-nil for each non-self-initiated transaction.

Reply:
Even if all current method users have squidPort, the method looks too dangerous to me: Other than the method implementation itself, there is no signal to the method caller that the method is only meant for requests that belong to a non-self-initiated transaction (and that know their master transaction).

Perhaps we can approach this from the other end? Self-initiated transactions are never intercepted so this method ought to return false for them, right? This logic, if valid, would allow us to convert/move the second asserted condition into a result condition.

As for the first asserted condition, a similar trick would only apply if we are certain that HttpRequests without a master transaction are never intercepted. I do not know whether that is true because I do not know what HttpRequest objects lack master transaction (if any). I am especially concerned that the method inspects _master_ transaction while the request object itself may have little to do with that transaction -- the request this object represents may not be intercepted even if the master transaction was (originated by a request on an intercepted connection). These concerns are a serious red flag for me, but I have not reviewed the entire update PR yet -- I am just responding to your comment.

If we cannot wiggle ourselves out of the first asserted condition, then we should probably move this method into MasterXaction where that condition would naturally become the responsibility of the caller.

Reply:
> certain that HttpRequests without a master transaction 

I wonder whether such HttpRequests exist at all: its both constructors assert on the passed master transaction. HttpRequest::masterXaction pointer may be nullified of course during the HttpRequest object lifetime - but this would look fairly strange.

> request object itself may have little to do with that transaction 

If you imply a Downloader request - it has its own MasterXaction object (which is probably wrong).

> we should probably move this method into MasterXaction

I considered this move - moreover, the caller code is responsible for assigning MasterXaction::squidPort (i.e., decide whether this transaction is not self-initiated). I created HttpRequest::interceptedXaction() for simplicity sake - to avoid request->masterXaction constructions in many places.

Reply:
> I wonder whether such HttpRequests exist at all: its both constructors assert on the passed master transaction. 

Great, the caller can rely on that, one way or another.

Right now, the first asserted condition can be misinterpreted as implying that it is the callers responsibility to check the masterXaction pointer. Based on your comments, it is not -- the condition is just a paranoid check of what ought to be true without any additional caller checks. If this method survives, this misdirection can be corrected using source code comments.

> HttpRequest::masterXaction pointer may be nullified of course during the HttpRequest object lifetime - but this would look fairly strange.

I assume you have checked that no such code exists today. I agree that it would probably be wrong for a request to suddenly lose its association with a master transaction. Given your statement about HttpRequest constructors, I would expect us to always maintain an "each HttpRequest knows its MasterXaction" invariant (that we should _use_ to our advantage).

> If you imply a Downloader request

I do not. My response was based on general principles, not current code/bugs.

> it has its own MasterXaction object (which is probably wrong).

IIRC, we have struggled with that decision. I do not recall specifics, but I would not be surprised if the current code represents a temporary ugly solution to work around some other problems. Right or wrong, I do not think that particular use case matters here as long as the principle itself stands -- a master transaction may be associated with many HttpRequest objects (which is what makes it a _master_ transaction in the first place).

> I created HttpRequest::interceptedXaction() for simplicity sake - to avoid request->masterXaction constructions in many places.

Yes, I understand the motivation. However, obscuring _what_ was intercepted cannot be justified by avoiding code duplication. We could, of course, add a masterWasIntercepted() or a similar method to solve this problem, but I think `r->masterXaction->intercepted()` or `r->master().intercepted()` would be overall better/clearer.